rust-runtime/aws-smithy-runtime/src/client/http/body/minimum_throughput/throughput.rs

```diff
@@ -97,90 +97,269 @@ impl From<(u64, Duration)> for Throughput {
     }
 }
 
-#[derive(Clone)]
+/// Represents a bin (or a cell) in a linear grid that represents a small chunk of time.
+#[derive(Copy, Clone, Debug)]
+enum Bin {
+    /// There is no data in this bin.
+    Empty,
+
+    /// No polling took place during this bin.
+    NoPolling,
+
+    /// This many bytes were transferred during this bin.
+    TransferredBytes(u64),
+
+    /// The user/remote was not providing/consuming data fast enough during this bin.
+    ///
+    /// The number is the number of bytes transferred, if this replaced TransferredBytes.
+    Pending(u64),
+}
+impl Bin {
+    fn merge(&mut self, other: Bin) {
+        use Bin::*;
+        // Assign values based on this priority order (highest priority higher up):
+        //   1. Pending
+        //   2. TransferredBytes
+        //   3. NoPolling
+        //   4. Empty
+        *self = match (other, *self) {
+            (Pending(other), this) => Pending(other + this.bytes()),
+            (TransferredBytes(other), this) => TransferredBytes(other + this.bytes()),
+            (other, NoPolling) => other,
+            (NoPolling, _) => panic!("can't merge NoPolling into bin"),
+            (other, Empty) => other,
+            (Empty, _) => panic!("can't merge Empty into bin"),
+        };
+    }
+
+    /// Number of bytes transferred during this bin
+    fn bytes(&self) -> u64 {
+        match *self {
+            Bin::Empty | Bin::NoPolling => 0,
+            Bin::TransferredBytes(bytes) | Bin::Pending(bytes) => bytes,
+        }
+    }
+}
+
+#[derive(Copy, Clone, Debug, Default)]
+struct BinCounts {
+    /// Number of bins with no data.
+    empty: usize,
+    /// Number of "no polling" bins.
+    no_polling: usize,
+    /// Number of "bytes transferred" bins.
+    transferred: usize,
+    /// Number of "pending" bins.
+    pending: usize,
+}
+
+/// Underlying stack-allocated linear grid buffer for tracking
+/// throughput events for [`ThroughputLogs`].
+#[derive(Copy, Clone, Debug)]
+struct LogBuffer<const N: usize> {
+    entries: [Bin; N],
+    // The length only needs to exist so that the `fill_gaps` function
+    // can differentiate between `Empty` due to there not having been enough
+    // time to establish a full buffer worth of data vs. `Empty` due to a
+    // polling gap. Once the length reaches N, it will never change again.
+    length: usize,
```

Comment:
what if you add a third bin option `Untouched` or something? Then we can remove this length logic which makes me a little uncomfortable due to off-by-one risk

Reply:
I had that approach originally, but it only moves the problem. Going that way, you have to track the time that the tracking started so that you know which cells can be flipped to NoPolling.